File: 04-json/examples.go
// Базовая работа с JSON в Go через encoding/json.
// Запуск: go run examples.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

// Тег `json:"name"` определяет имя поля в JSON.
// `omitempty` пропускает поле, если значение нулевое.
type User struct {
	Name    string   `json:"name"`
	Age     int      `json:"age"`
	Skills  []string `json:"skills"`
	Address *string  `json:"address"` // указатель → null когда nil
}

func main() {
	// ── Сериализация ───────────────────────────────────────────────
	u := User{
		Name:    "Ден",
		Age:     29,
		Skills:  []string{"python", "go"},
		Address: nil,
	}

	b, err := json.Marshal(u)
	if err != nil {
		panic(err)
	}
	fmt.Println("compact:", string(b))

	// С отступами
	pretty, _ := json.MarshalIndent(u, "", "  ")
	fmt.Println("pretty:")
	fmt.Println(string(pretty))

	// ── Десериализация в struct ────────────────────────────────────
	raw := []byte(`{"name":"Ден","age":29,"skills":["python","go"],"address":null}`)
	var parsed User
	if err := json.Unmarshal(raw, &parsed); err != nil {
		panic(err)
	}
	fmt.Printf("parsed struct: %+v\n", parsed)

	// ── Десериализация в map (когда схема неизвестна) ──────────────
	var m map[string]any
	_ = json.Unmarshal(raw, &m)
	fmt.Printf("as map: %v\n", m)
	// внимание: числа в map[string]any парсятся как float64
	if age, ok := m["age"].(float64); ok {
		fmt.Println("age как float64:", age)
	}

	// ── Запись / чтение файла ──────────────────────────────────────
	f, _ := os.Create("user.json")
	enc := json.NewEncoder(f)
	enc.SetIndent("", "  ")
	_ = enc.Encode(u)
	_ = f.Close()

	f2, _ := os.Open("user.json")
	defer f2.Close()
	var loaded User
	_ = json.NewDecoder(f2).Decode(&loaded)
	fmt.Printf("from file: %+v\n", loaded)

	// ── Обработка ошибок ───────────────────────────────────────────
	if err := json.Unmarshal([]byte(`{"name": broken}`), &parsed); err != nil {
		fmt.Println("ошибка парсинга:", err)
	}
}

File: 04-json/examples.py
"""
Базовая работа с JSON в Python через стандартный модуль json.
Запуск: python examples.py
"""

import json
from dataclasses import dataclass, asdict


# ── Сериализация: объект → строка ────────────────────────────────────
data = {
    "name": "Ден",
    "age": 29,
    "skills": ["python", "go"],
    "active": True,
    "address": None,
}

# Минимально (одна строка)
s = json.dumps(data)
print("compact:", s)

# Красиво и без \uXXXX
pretty = json.dumps(data, ensure_ascii=False, indent=2)
print("pretty:\n" + pretty)


# ── Десериализация: строка → объект ──────────────────────────────────
raw = '{"name": "Ден", "age": 29, "skills": ["python", "go"]}'
obj = json.loads(raw)
print("parsed:", obj, type(obj))


# ── Чтение / запись файла ────────────────────────────────────────────
with open("user.json", "w", encoding="utf-8") as f:
    json.dump(data, f, ensure_ascii=False, indent=2)

with open("user.json", "r", encoding="utf-8") as f:
    loaded = json.load(f)
print("from file:", loaded)


# ── Кастомные классы через dataclass ─────────────────────────────────
@dataclass
class User:
    name: str
    age: int
    skills: list[str]


u = User(name="Ден", age=29, skills=["python", "go"])
encoded = json.dumps(asdict(u), ensure_ascii=False)
print("dataclass:", encoded)

decoded_dict = json.loads(encoded)
u2 = User(**decoded_dict)
print("back to obj:", u2)


# ── Не сериализуемые объекты ─────────────────────────────────────────
import datetime


def default(o: object) -> str:
    if isinstance(o, datetime.datetime):
        return o.isoformat()
    raise TypeError(f"{type(o)} is not JSON serializable")


now = {"ts": datetime.datetime(2026, 5, 6, 12, 0)}
print("with custom default:", json.dumps(now, default=default))


# ── Обработка ошибок ─────────────────────────────────────────────────
try:
    json.loads("{name: 'broken'}")
except json.JSONDecodeError as e:
    print(f"невалидный JSON: {e.msg} в позиции {e.pos}")
